react/src/data/dependency.js

```diff
@@ -0,0 +1,69 @@
+import { useSelector, useDispatch, useStore } from 'react-redux';
```

Comment:
I rewrote this to show a few things that we do here differently. 

We name objects by what they are, not how they are used. An SPath is the thing. A Dependency is how it is used. It could be used for other things. 

Fail fast is a deep philosophy. Unless there is a good everything should be fail fast. This surfaces defects instead of burying them.

I'm not sure about the names, but I tired to express what I thought they were. The point is intent, not prescription. 

We don't put blank lines in a function. That usually indicates another function is needed that describes the block of code by its name.

Temporary variables are avoided. They introduce cognitive load. "Why does this name mean?" "Where is it referenced?" "What are the implicit couplings and dependencies?" And so on.

The single letter variable naming system is to make that apparent. More than three single letter variables indicates modularization is required.

It took me quite a while to figure out this code. The long names hindered my ability to understand, because I needed to understand what the "hook" part of it. However, that's not what it does. It's about mapping fields to values. The algorithm is important for Redux to reduce dependencies, but that should be described by the name where it is used. HookFieldMap gives enough of a clue that this is about hooks. FieldMap is about what this code does.

To that end, I turned it into a map so filtering didn't need to happen on lookups. That code was confusing:

```js
    getHookedDependency = (dependency) => {
        return this.hookedDependencies.find(hookedDependency => {
            return (dependency.modelName === hookedDependency.modelName &&
                dependency.fieldName === hookedDependency.fieldName);
        })
    }
```

By making SPath (below) responsible for storing the fully qualified name (spath) and the decomposed name (modelName and fieldName), the method was trivialized, and also surfaced the problem that it wasn't fail fast. If the model is not found, the above code will return None. 

I did make the assumption that fields wanted to be reduced, too. This could be wrong. If so, this should be called out in the code clearly: preserving duplicates.

I don't understand modern javascript well enough, but I think the `...` and `{}` "expansions" simplified the code. Again, this eliminates many unnecessary temporary variables.

There are other [DesignHints](https://github.com/radiasoft/pykern/wiki/DesignHints) here. Please read through the code and discuss with me or in this PR.

```js
export class SPath {
    constructor(spath) {
        {this.modelName, this.fieldName} = spath.match(/^(\w[^\.]*))\.(\w[^\.]*)$/);
        if (!this.modelName ){
            throw new Error(`spath=${spath} must be in model.field form`);
        }
        this.spath = spath;
    };
};

export class HookFieldMap {
    constructor({ dependencies, models, schemaModels }) {
        let cache = {};
        let modelInfo = (name) => {
            if (! (name in cache)) {
                //TODO: is this a fail fast object?
                let m = models.forModel(name);
                cache[name] = {
                    ...m,
                    modelSchema: schemaModels[name],
                    modelValue: m.hookModel(),
                };
            }
            return cache[name];
        };
        this.fields = Object.fromEntries(
            dependencies.map(
                (spath) => {
                    let m = modelInfo(spath.modelName);
                    let s = m.modelSchema[spath.fieldName];
                    if(! s) {
                        throw new Error(`field=${path.field} not in model=${spath.model} schema=${m.modelSchema}`)
                    }
                    return [
                        spath,
                        {
                            ...spath,
                            ...s,
                            model: m,
                            value: m.value[spath.fieldName],
                        },
                    ],
                },
            ),
        );
    };

    find = (spath) => {
        //TODO: needs to be a fail fast object
        if (! (spath in this.fields) {
            throw new Error(`field=${spath} not found in HookFieldMap`);
        }
        return this.fields[spath];
    };
}
```